sourcing-new/src/main/scala/ch/epfl/bluebrain/nexus/sourcingnew/projections/ProjectionStream.scala

```diff
@@ -0,0 +1,203 @@
+        message match {
+          case s: SuccessMessage[A] =>
+            fetchResource(s.value)
+              .map {
+                _.flatMap(collect) match {
+                  case Some(value) =>
+                    message.map(_ => value)
+                  case _              => s.discarded
+                }
+              }
+              .recoverWith {
+                //TODO Handle errors properly
+                case NonFatal(err) => F.pure(s.failed(err))
+              }
+          case e: SkippedMessage      => F.pure(e)
+        }
+  }
+
+  implicit class SimpleStreamOps[F[_]: Timer, A: Encoder]
+    (val stream: Stream[F, Message[A]])(implicit override val F: ConcurrentEffect[F]) extends StreamOps[F, A] {
+
+    type O[_] = Message[_]
+
+    def resource[B, R](fetchResource: A => F[Option[R]], collect: R => Option[B]): Stream[F, Message[B]] =
+      stream.mapAsync(1)(toResource(fetchResource, collect))
+
+    /**
+      * Fetch a resource without transformation
+      * @param fetchResource
+      * @tparam R
+      * @return
+      */
+    def resourceIdentity[R](fetchResource: A => F[Option[R]]): Stream[F, Message[R]] =
+      resource(fetchResource, (r:R) => Option(r))
+
+    /**
+      * Fetch a resource and maps and filter thanks to a partial function
+      * @param fetchResource
+      * @param collect
+      * @tparam B
+      * @tparam R
+      * @return
+      */
+    def resourceCollect[B, R](fetchResource: A => F[Option[R]], collect: PartialFunction[R, B]): Stream[F, O[B]] =
+      resource(fetchResource, collect.lift)
+
+    /**
+      * Apply the given function for every success message
+      *
+      * If the function gives an error, the message will be marked as failed,
+      * It will remain unmodified otherwise
+      *
+      * @param f
+      * @return
+      */
+    def runAsync(f: A => F[Unit]): Stream[F, Message[A]] =
+      stream.mapAsync(1) {  message: Message[A] =>
+        message match {
+          case s: SuccessMessage[A] =>
+            (f(s.value) >> F.pure[Message[A]](s))
+              .recoverWith {
+                err => F.pure(s.failed(err))
+              }
+          case v => F.pure(v)
+        }
+      }
+
+    def persistErrors(persistErrors: ErrorMessage => F[Unit]): Stream[F, Message[A]] =
+      stream.mapAsync(1) { message =>
+        message match {
+          case e: ErrorMessage => persistErrors(e) >> F.pure(message)
+          case _: Message[A] => F.pure(message)
+        }
+      }
+
+    //def toProgress(initial: ProjectionProgress = NoProgress): Stream[F, ProjectionProgress]
+
+  }
+
+  implicit class ChunckStreamOps[F[_]: Timer, A: Encoder]
+    (val stream: Stream[F, Chunk[Message[A]]])(implicit override val F: ConcurrentEffect[F]) extends StreamOps[F, A] {
+
+    type O[_] = Chunk[Message[_]]
+
+    private def discardDuplicates(chunk: Chunk[Message[A]]): List[Message[A]] = {
+      chunk.toList.foldRight((Set.empty[String], List.empty[Message[A]])) {
```

Comment:
Isn't foldRight not as efficient as foldLeft and can cause Stackoverflow?

This same code can be implemented as foldLeft but instead of discarding the new messages, discarding the old ones overriding the new ones

Reply:
foldRight on Scala collections is reverse + foldLeft, so it is ok.
There is the cost of the reverse but I don't think we will have so large chunks